### PR TITLE
refactor(ivy): validate that identifier identity in emitted output

### PR DESCRIPTION
**Pull request by chuckjaz**
Modifies validation syntax to generate back references to ensure
that identifiers are used consistently.

Introduced … to allow validating constant definition and usage.

## PR Type
What kind of change does this PR introduce?

```
[x] Refactoring (no functional changes, no api changes)
```

## What is the new behavior?

Modifies validation syntax to generate back references to ensure
that identifiers are used consistently.

Introduced … to allow validating constant definition and usage.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by vicb — CHANGES_REQUESTED** [review]
Great change!

As discussed earlier I would love to see `mock_compile_spec.ts` if I have to debug/enhance the test code at some point. At the very least add a one liner comment to the fns and on each of the important blocks ?

What about replacing `…` by something easier to type ie `// ...` ?

File: packages/compiler/test/render3/mock_compile.ts [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    }
+    fail(
+        `Test helper failure: Expected expression failed but the reporting logic could not find where it failed in: ${source}`);
+  }
+}
+
+const IDENT_LIKE = /^[a-z][A-Z]/;
+const SPECIAL_RE_CHAR = /\/|\(|\)|\||\*|\+|\[|\]|\{|\}|\$/g;
+function r(...pieces: (string | RegExp)[]): RegExp {
+  let results: string[] = [];
```
Comment (vicb) [inline packages/compiler/test/render3/mock_compile.ts]: didn't lint use to enforce `const` ?

**Comment by chuckjaz** [comment]
Added unit tests.

Explained in comments why … is a good thing ;-).

**Review by vicb — APPROVED** [review]
I love it !

**Pull request closed without merging**